pkg/snapshotter/snapshotter.go

```diff
@@ -44,26 +49,61 @@ type Snapshotter interface {
 }
 
 type snapshot struct {
-	conn *grpc.ClientConn
+	conn    *grpc.ClientConn
+	backoff *wait.Backoff
 }
 
+// SnapshottingState is the state of volume snapshotting. It tells the controller if snapshotting could
+// be in progress in the background after a CreateSnapshotExt() call returns or the snapshotting is
+// 100% finished (with or without success)
+type SnapshottingState string
+
+const (
+	// SnapshottingInBackground tells the controller that snapshotting may be in progress
+	// after CreateSnapshotExt exits.
+	SnapshottingInBackground SnapshottingState = "Background"
+	// SnapshottingFinished tells the controller that snapshotting is not running in the background
+	// and has exited successfully or with errors.
+	SnapshottingFinished SnapshottingState = "Finishing"
+	// SnapshottingNoChange tells the controller that snapshotting status has not changed since
+	// CreateSnapshotExt was called.
+	SnapshottingNoChange SnapshottingState = "NoChange"
+
+	// createSnapshotRetryCount is the number of retries that will be attempted
+	// during snapshot creation
+	createSnapshotRetryCount = 5
+
+	// createSnasphotRetryInterval is the interval between snapshot creation retries
+	createSnapshotRetryInterval = 10 * time.Second
+)
+
+var (
+	createSnapshotBackoff *wait.Backoff
+)
+
 func NewSnapshotter(conn *grpc.ClientConn) Snapshotter {
 	return &snapshot{
 		conn: conn,
+		backoff: &wait.Backoff{
+			Duration: createSnapshotRetryInterval,
+			Factor:   1, //linear backoff
+			Steps:    createSnapshotRetryCount,
+			Cap:      createSnapshotRetryInterval,
+		},
 	}
 }
 
-func (s *snapshot) CreateSnapshot(ctx context.Context, snapshotName string, volume *v1.PersistentVolume, parameters map[string]string, snapshotterCredentials map[string]string) (string, string, time.Time, int64, bool, error) {
+func (s *snapshot) CreateSnapshot(ctx context.Context, snapshotName string, volume *v1.PersistentVolume, parameters map[string]string, snapshotterCredentials map[string]string) (string, string, time.Time, int64, bool, SnapshottingState, error) {
 	klog.V(5).Infof("CSI CreateSnapshot: %s", snapshotName)
 	if volume.Spec.CSI == nil {
-		return "", "", time.Time{}, 0, false, fmt.Errorf("CSIPersistentVolumeSource not defined in spec")
+		return "", "", time.Time{}, 0, false, SnapshottingFinished, fmt.Errorf("CSIPersistentVolumeSource not defined in spec")
 	}
 
 	client := csi.NewControllerClient(s.conn)
 
 	driverName, err := csirpc.GetDriverName(ctx, s.conn)
```

Comment:
Consider caching the driver name (in a separate PR), it should not change during snapshotter lifetime.

Reply:
Added an issue to fix this in a separate PR.